### PR TITLE
Initial restructuring of configuration loading mechanism

### PR DESCRIPTION
**Pull request by jessemyers**
Microcosm has always had a `loader` abstraction for loading configuration
from a specific source; however, the overall mechanism of loading configuration
from defaults, merging in loaders, etc. has been inline, during object graph
creation. Now we have (the beginnnings) of an abstraction layer.

### CODE REVIEW

File: microcosm/config/api.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+def get_defaults(func):
+    """
+    Retrieve the defaults for a factory function.
+
+    """
+    return getattr(func, DEFAULTS, {})
+
+
+def configure(registry, metadata, loader):
```
Comment (jessemyers) [inline microcosm/config/api.py]: If you _squint_ you should notice that `configure` and `configure_scoped` have the same structure. My next step is to build better abstraction around these operations.

File: microcosm/config/model.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+"""Configuration model and loading"""
```
Comment (jessemyers) [inline microcosm/config/model.py]: Moved into `microcosm/config`

File: microcosm/configuration.py [hunk 3]
```diff
@@ -1,65 +1,2 @@
...
-                self[key].merge(value)
-            elif isinstance(value, list) and isinstance(self.get(key), list):
-                # append
-                self[key] += value
-            else:
-                # set the new value
-                self[key] = value
-
-    __setitem__ = __setattr__
+# backwards compatbility import
```
Comment (jessemyers) [inline microcosm/configuration.py]: Temporary.

File: microcosm/config/api.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    })
+    config.merge(loader(metadata))
+    return config
+
+
+def configure_scoped(graph, key, factory, loader):
+    """
+    Build a sub-configuration for a specific binding.
+
+    :params key: a binding key
```
Comment (cdm-ium) [inline microcosm/config/api.py]: :D I think there is more than just a param for which key to bind to.

**Pull request closed without merging**